src/main/serialport.ts

```diff
@@ -0,0 +1,348 @@
+    if (!port) {
+        return false;
+    }
+
+    return port.isOpen;
+};
+
+/* Return true if port is closed, false otherwise */
+export const closeSerialPort = (
+    path: string,
+    sender: Renderer
+): [Error | undefined, boolean] => {
+    const port = serialPorts.get(path)?.serialPort;
+    let portWasClosed = false;
+    if (!port) {
+        return [
+            Error(
+                `Could not find SerialPort with path ${path}, and could not be closed.`
+            ),
+            portWasClosed,
+        ];
+    }
+
+    const renderers = serialPorts.get(path)?.renderers;
+    if (renderers) {
+        // FIXME: Issue is I can't get the potential Error from port.close().
+        // removeRenderer will close the port and return true if list of renderers is empty after removing sender.
+        portWasClosed = removeRenderer(path, sender);
+    }
+    return [undefined, portWasClosed];
+};
+
+const removeRenderer = (path: string, sender: Renderer): boolean => {
+    let closedPort = false;
+    const openPort = serialPorts.get(path);
+    if (openPort) {
+        openPort.renderers = openPort.renderers.filter(
+            renderer => renderer.id !== sender.id
+        );
+
+        if (openPort.renderers.length === 0) {
+            if (openPort.serialPort.isOpen) {
+                openPort.serialPort.close();
+                serialPorts.delete(path);
+                closedPort = true;
+            }
+        } else {
+            serialPorts.set(path, openPort);
+        }
+    }
+    return closedPort;
+};
+
+export const update = (path: string, baudRate: number) => {
+    const openPort = serialPorts.get(path);
+
+    if (!openPort) {
+        return;
+    }
+
+    const { serialPort, renderers } = openPort;
+
+    return new Promise<void>((resolve, reject) => {
+        serialPort.update({ baudRate }, error => {
+            if (error) {
+                reject(error);
+            }
+            renderers.forEach(renderer => {
+                renderer.send(
+                    SERIALPORT_CHANNEL.ON_UPDATE,
+                    serialPort.settings
+                );
+            });
+            resolve();
+        });
+    });
+};
+
+export const set = (path: string, newOptions: SetOptions) => {
+    const openPort = serialPorts.get(path);
+
+    if (!openPort) {
+        return;
+    }
+
+    const { serialPort, renderers } = openPort;
+
+    serialPort.set(newOptions);
+
+    renderers.forEach(renderer => {
+        // TODO: Review if we actually want to send newOptions?
+        // You also have a get() function, but it's only a subset of SetOptions, called PortStatus
+        renderer.send(SERIALPORT_CHANNEL.ON_SET, newOptions);
+    });
+};
+
+const openNewSerialPort = (options: SerialPortOpenOptions<AutoDetectTypes>) => {
+    const { path } = options;
+    const openPort = serialPorts.get(path);
+    if (openPort && openPort.serialPort.isOpen) {
+        openPort.serialPort.close();
```

Comment:
This will fails in some cases as the port will be closing and if you try to reopen without waiting for the on closed callback you will get Access Denied on windows at least...

Consider adding the code from 292 till 324 in a callback passed to closed if port is open and if not just manually call this callback

Reply:
Fixed in commit #16fa5e9.